apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,438 @@
+   *
+   * @public
+   */
+  connected: false,
+
+  /**
+   * Debug message.
+   *
+   * @type {Boolean} turn on/off the console log
+   */
+  onDebug: false,
+
+  /**
+   * Store a reference of the default adapter.
+   *
+   * @private
+   */
+  _adapter: null,
+
+  /**
+   * Hold instance of bluetooth attribute change handler.
+   *
+   * @private
+   */
+  _defaultAdapterChangeHandler: null,
+
+  /**
+   * Hold instance of bluetooth adapter state change handler.
+   *
+   * @private
+   */
+  _stateChangeHandler: null,
+
+  /**
+   * Hold instance of bluetooth attribute change handler in
+   * getAdapter when the default adapter is not available.
+   *
+   * @private
+   */
+  _promiseAdapterChangeHandler: null,
+
+  /**
+   * Hold instance of the getAdapter promise
+   * when the default adapter is not available.
+   */
+  _cacheGetAdapterPromise: null,
+
+  /**
+   * Store a reference of the Bluetooth API.
+   *
+   * @private
+   */
+  _bluetooth: null,
+
+  /**
+   * Store a reference of the Bluetooth API.
+   *
+   * @private
+   */
+  _isEnabled: false,
+
+  /**
+   * Build-in Bluetooth profiles
+   *
+   * @public
+   */
+  get Profiles() {
+    return {
+      HFP: 'hfp',   // Hands-Free Profile
+      OPP: 'opp',   // Object Push Profile
+      A2DP: 'a2dp', // A2DP status
+      SCO: 'sco'    // Synchronous Connection-Oriented
+    };
+  },
+
+  /**
+   * Set profile connect state.
+   *
+   * @public
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      // if (profile === 'opp' && this.transferIcon) {
+      //   this.transferIcon.update();
+      // }
+      // if (profile === 'a2dp' && this.headphoneIcon) {
+      //   this.headphoneIcon.update();
+      // }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+    SettingsListener.observe('bluetooth.enabled', true, (value) => {
+      if (!this._bluetooth) {
```

Comment:
I think we can remove the entire handler in v2 because there should be no one writes to `bluetooth.enabled` directly except for settings app and the module itself.